### DISCUSSION TITLE
Memory consumed

### DISCUSSION THREAD

**Discussion by perceptualrobots** [cyberbotics/wrestling#75] [Q&A]
When I run multiple instances of webots I am finding all memory gets consumed. The processes then either grind to a halt or get an OS access violation and crash.

![Screenshot 2023-05-01 135943](https://user-images.githubusercontent.com/67976964/235518263-15f3c9e6-d4d6-4663-b74c-59dc4cbdd634.png)
![Screenshot 2023-05-01 140007](https://user-images.githubusercontent.com/67976964/235518282-247ee62c-f8a8-4fc6-90f3-d987fd600585.png)

I run the instances like this. 

`"C:\Program Files\Webots\msys64\mingw64\bin\webotsw.exe" --port=1234 --no-rendering --batch --mode=fast  --stdout  --stderr wrestling-nosync.wbt`

The wbt has synchronization FALSE. I am using extern controllers running hundreds of times.

Am I doing something wrong?

**Comment by perceptualrobots** [comment]
Actually, multiple instances is irrelevant, it happens with just one instance.

**Comment by omichel** [comment]
That looks very strange. Are you sure your robot control program is not responsible for eating up all the memory? If you open the ">" sign on `webots-bin.exe (3)` and `webots-bin.exe (2)` (see your screenshot), you should be able to see more details on who is consuming what.

Also, if you run a single instance of Webots, it could be that there remain some zombie processes of Webots still consuming some memory resources.

Finally, if the Webots simulation is running for a very long time (you said using extern controllers running hundreds of time), it could be that there is some small memory leak in Webots causing the memory used by Webots to grow indefinitely. Which version of Webots are you using? Is that a recent nightly build of Webots R2023a-rev1?

Although Webots is designed to run several instances in parallel, there is a limit on the number of instances you can launch on a single computer. This limit depends on both the complexity of the simulation and the horsepower of the computer (mainly CPU, GPU, RAM).

**Reply by perceptualrobots** [reply]
> That looks very strange. Are you sure your robot control program is not responsible for eating up all the memory? If you open the ">" sign on `webots-bin.exe (3)` and `webots-bin.exe (2)` (see your screenshot), you should be able to see more details on who is consuming what.

It seems to be the .exe.

![Screenshot 2023-05-02 115031](https://user-images.githubusercontent.com/67976964/235648532-9cb9fb7b-c0f4-4156-a32c-fceebe2c0d56.png)

> Also, if you run a single instance of Webots, it could be that there remain some zombie processes of Webots still consuming some memory resources.

I get the issue even if I restart the machine.

> Finally, if the Webots simulation is running for a very long time (you said using extern controllers running hundreds of time), it could be that there is some small memory leak in Webots causing the memory used by Webots to grow indefinitely. Which version of Webots are you using? Is that a recent nightly build of Webots R2023a-rev1?

It's R2023a-rev1 from Apr, 25.

> Although Webots is designed to run several instances in parallel, there is a limit on the number of instances you can launch on a single computer. This limit depends on both the complexity of the simulation and the horsepower of the computer (mainly CPU, GPU, RAM).

It was actually working last week, running for over 24 hours without problem (I didn't notice any memory issues). I commincate with participant.py (robot supervisor) by TCPIP from my python controller. On participant.py there is a TCPIP listener server and a client on my controller.

I have made some changes over the last week, though I don't see how they would affect the webots.exe process as the communication between participant and webots hasn't changed (I read and write motor positions). Participant was running with a fixed game duration (10 secs). One thing I have change is that the controller now sends the game duration to participant.py so I have better control over that variable. The main other thing I have changed is that I was closing the TCPIP connection after each run (and simulationReset) of the wrestler, now I keep the TCPIP connection open for the whole session.

Do you have any ideas of how changes to my two python processes could have affected the webots.exe process? Is the participant extern sending more than just the motor commands over TCPIP to webots?

**Reply by omichel** [reply]
No, I don't have any idea on what could have caused this memory leak. I would recommend you to investigate by yourself, e.g., start from the version of your controller which wasn't causing any memory leak, and start adding features one-by-one towards the new version which is causing the memory leak. If after adding some specific feature, you start to see the memory leak, then you will have identified the culprit. Please let me know about it so that we can further investigate and hopefully fix it.

**Reply by perceptualrobots** [reply]
I have figured out that the change that resulted in the problem was using a single TCP connection.

My main loop is this:
```
 wrestler = WrestlerSupervisorServer()
        while True:
            wrestler.simulationReset()
            wrestler.run(port=port)

```
Previously, within each run I would accept a connection and then close it at the end. The change involved keeping a connection open continuously, with a singleton class, for all calls to run.

Do you know a reason why using a singleton would result in the problem? Or why having a connection open continuously would be a problem? (I may try without using a singleton).

Here's the Server code I am using from participant.py.

```
class Server():
    def __init__(self, host='localhost', port=6666 , buf_size=1024):
        self.buf_size = buf_size
        self.host=host
        self.port=port
        self.listen()
            
    def listen(self):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.bind((self.host, self.port))
            s.listen()
            print(f"Waiting for connection on {self.host}:{self.port}")
            self.connection, addr = s.accept()
            print(f"Connected by {addr}")
        self.open=True
        
        
    def get(self):
        data = self.connection.recv(self.buf_size)
        return data.decode()
    
    def put(self, data):
        d = str.encode(data)
        self.connection.sendall(d)

    def get_dict(self):
        length=-1
        lostconnection=False
        try:
            data = self.connection.recv(self.buf_size)
            length = len(data)
        except:
            lostconnection=True
            
        if lostconnection or length==0:
            self.listen()
            data = self.connection.recv(self.buf_size)
            
        dict = eval(data.decode())
        #print('recv',dict)
        return dict
    
    
    def put_dict(self, dict):
        #print('send',dict)       
        json_object = json.dumps(dict) 
        d = str.encode(json_object)
        self.connection.sendall(d)

    def isOpen(self):
        return self.open

    def finish(self):
        self.open = False

    def close(self):
        self.open=False
        self.connection.close()
        print('Closed server connection')
```
```
class ServerConnectionManager:
    "A utility for managing a client socket connection."
    __instance = None
    @staticmethod 
    def getInstance():
        """ Static access method. """
        if ServerConnectionManager.__instance == None:
             ServerConnectionManager()
        return ServerConnectionManager.__instance
    
    def __init__(self, host='localhost', port=6666 , buf_size=1024):
        """ Virtually private constructor. """
        if ServerConnectionManager.__instance != None:
             raise Exception("This class is a singleton!")
        else:
             self.host='localhost'
             self.port=6666
             self.buf_size=1024
             self.connected=False
             ServerConnectionManager.__instance = self

    def connect(self):
        self.server = Server(host=self.host, port=self.port, buf_size=self.buf_size)
        self.connected=self.server.isOpen()

    def isOpen(self):
        return self.connected
    
    def close(self):
        self.server.close()
        self.connected=self.server.isOpen()
                 
    def send(self, data):
        if not self.isOpen():
            self.connect()
                
        self.server.put_dict(data)

    def receive(self):
        if not self.isOpen():
            self.connect()

        recv = self.server.get_dict()
        return recv            
    
    def set_port(self, port):
        self.port=port
        
    def set_host(self, host):
        self.host=host        
```

**Comment by perceptualrobots** [comment]
On further testing:

On one machine I have one webots running which remains stable at around 0.6 Gb RAM.

On another machine I had four running and the memory was increasing though at a slower rate than previously. The RAM for each got up to around 9, 5, 4 and 1 Gb. I killed the largest two and the others seem stable. I will leave them overnight and see how they go.

**Comment by perceptualrobots** [comment]
The issue persisted overnight. I'd reduced to one instance on each machine and during the night webots crashed on both.  See windows event viewer below.

I think reverting to multiple TCP connections didn't resolve anything, it just slowed things down so the memory issue was slower.

Could it be due to garbage collection? I am running hundreds of games, each of which may only last less than a second. Maybe this is too fast for the garbage collection to keep up with?


```
Log Name:      Application
Source:        Application Hang
Date:          02/05/2023 22:33:26
Event ID:      1002
Task Category: (101)
Level:         Error
Keywords:      Classic
User:          N/A
Computer:      DESKTOP-5O07H5P
Description:
The program webots-bin.exe version 0.0.0.0 stopped interacting with Windows and was closed. To see if more information about the problem is available, check the problem history in the Security and Maintenance control panel.
 Process ID: 3540
 Start Time: 01d97d31fe78e31f
 Termination Time: 4294967295
 Application Path: C:\Program Files\Webots\msys64\mingw64\bin\webots-bin.exe
 Report Id: 1773152e-6fb1-4fd9-a91a-8465f745db59
 Faulting package full name: 
 Faulting package-relative application ID: 
 Hang type: Unknown

Event Xml:
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
  <System>
    <Provider Name="Application Hang" />
    <EventID Qualifiers="0">1002</EventID>
    <Version>0</Version>
    <Level>2</Level>
    <Task>101</Task>
    <Opcode>0</Opcode>
    <Keywords>0x80000000000000</Keywords>
    <TimeCreated SystemTime="2023-05-02T21:33:26.3645900Z" />
    <EventRecordID>32536</EventRecordID>
    <Correlation />
    <Execution ProcessID="0" ThreadID="0" />
    <Channel>Application</Channel>
    <Computer>DESKTOP-5O07H5P</Computer>
    <Security />
  </System>
  <EventData>
    <Data>webots-bin.exe</Data>
    <Data>0.0.0.0</Data>
    <Data>3540</Data>
    <Data>01d97d31fe78e31f</Data>
    <Data>4294967295</Data>
    <Data>C:\Program Files\Webots\msys64\mingw64\bin\webots-bin.exe</Data>
    <Data>1773152e-6fb1-4fd9-a91a-8465f745db59</Data>
    <Data>
    </Data>
    <Data>
    </Data>
    <Data>Unknown</Data>
    <Binary>55006E006B006E006F0077006E0000000000</Binary>
  </EventData>
</Event>
```

**Reply by omichel** [reply]
That's probably a memory leak in Webots or in the controller library. A way to work around the problem is to modify the referee supervisor so that it [quits](https://cyberbotics.com/doc/reference/supervisor#wb_supervisor_simulation_quit) when done and have a script that launches Webots and repeatedly re-launches Webots as soon as it quits. This way, the memory is cleaned-up regularly and shouldn't grow until Webots crashes.

**Reply by perceptualrobots** [reply]
> That's probably a memory leak in Webots or in the controller library. A way to work around the problem is to modify the referee supervisor so that it [quits](https://cyberbotics.com/doc/reference/supervisor#wb_supervisor_simulation_quit) when done and have a script that launches Webots and repeatedly re-launches Webots as soon as it quits. This way, the memory is cleaned-up regularly and shouldn't grow until Webots crashes.

Ah, a couple of weeks ago I was looking to control webots from python/bat script but couldn't work out how to exit, as this seems to do.

I am runnng a population of 1000 for 25 generations, so probably don't want to start/stop it 25,000, though maybe I can do it intermittently based upon system memory.

**Reply by perceptualrobots** [reply]
I managed to reproduce the issue with the 'simplified' code below. However, it is not consistent.

I have two machines, desktop and laptop. My latest code, for running my GA, had the memory problem on both.

So, I reverted my code to the 26th April, when I know my GA was working and ran for 36 hours.

However, on the laptop with this previous code I still had the memory problem. I tried this previous code on my desktop and did not have the memory problem.

But running the simplified code on my laptop did have the memory problem, with RAM usage increasing at a rate of about 2Gb per minute. I wondered if the problem was due to things I was running on the laptop. So, I restarted the laptop and have not seen the problem. Could there be some negative interaction with other, graphic, applications? The sort of applications I was running were WSL, Thunderbird, VScode, webots and Amazon Prime viseo in  Google Chrome.

Maybe this simplified will help you investigate? In the meantime, I will try restarting webots from with my code.

```
participant/participant.py
class WrestlerSupervisorMemory(Supervisor):
    def initSupervisor(self):
        self.robot = [0] * 2
        self.robot[1] = self.getFromDef('WRESTLER_BLUE').getFromProtoDef('HEAD_SLOT')
        self.robot[0] = self.getFromDef('WRESTLER_RED').getFromProtoDef('HEAD_SLOT')
        self.min = [[0] * 3 for i in range(2)]
        self.max = [[0] * 3 for i in range(2)]
        for i in range(2):
            self.min[i] = self.robot[i].getPosition()
            self.max[i] = self.robot[i].getPosition()
        self.coverage = [0] * 2
        self.ko_count = [0] * 2

    def run(self, CI, time_step=None, max_loops=None):
        self.rr = RobotAccess(self)
        self.initSupervisor()
        self.motion_library = MotionLibrary()
        self.time_step = int(self.getBasicTimeStep())
        ko_labels = ['', '']
        coverage_labels = ['', '']
        # Performance output used by automated CI script
        game_duration = 100 # 5000 # 3 * 60 * 1000  # a game lasts 3 minutes
        # retrieves the WorldInfo.basicTimeTime (ms) from the world file
        fileTimeStep=int(self.getBasicTimeStep())
        if time_step==None:
            time_step = fileTimeStep
        #print(time_step)
        
        if max_loops==None:
            max_loops = game_duration/time_step
        
        time = 0
        seconds = -1
        ko = -1
        loops=0
        initial={'LHipPitch': 0.0, 'LKneePitch': 0.0, 'LAnklePitch': 0.0, 'RHipPitch': 0.0, 'RKneePitch': 0.0, 'RAnklePitch': 0.0}
        
        while self.step(self.time_step) != -1:  # mandatory function to make the simulation run
            # if time > 22000:
            #     self.motion_library.play('Backwards')
            # else:
            #     self.motion_library.play('Forwards')
                
            legs = self.rr.read()
            # print(legs)         
            actions = {'LHipPitch': 0.0, 'LKneePitch': 0.0, 'LAnklePitch': 0.0, 'RHipPitch': 0.0, 'RKneePitch': 0.0, 'RAnklePitch': 0.0}
            self.rr.set(initial, actions)   
            
            ko = self.evaluation(time, seconds, ko_labels, coverage_labels, ko)            

            # if self.step(time_step) == -1 or time > game_duration or ko != -1 or loops==max_loops:
            if ko != -1 or loops==max_loops:
                break
            time += time_step
            loops = loops+1
            
        print(f'Time={time} loops={loops} time_step={time_step} ftime_step={fileTimeStep}')
        if ko == 0:
            print('Red is KO. Blue wins!')
            performance = 0
        elif ko == 1:
            print('Blue is KO. Red wins!')
            performance = 1
        # in case of coverage equality, blue wins
        elif self.coverage[0] > self.coverage[1]:
            print('Red wins coverage: %s > %s' % (self.coverage[0], self.coverage[1]))
            performance = 1
        else:
            print('Blue wins coverage: %s >= %s' % (self.coverage[1], self.coverage[0]))
            performance = 0

    def evaluation(self, time, seconds, ko_labels, coverage_labels, ko):
        if time % 200 == 0:
            s = int(time / 1000) % 60
            if seconds != s:
                seconds = s
                minutes = int(time / 60000)
                #print(f'{time} {minutes:02}:{seconds:02}')
            box = [0] * 3
            for i in range(2):
                position = self.robot[i].getPosition()
                color = 0xff0000 if i == 0 else 0x0000ff
                if abs(position[0]) < 1 and abs(position[1]) < 1:  # inside the ring
                    coverage = 0
                    for j in range(2):
                        if position[j] < self.min[i][j]:
                            self.min[i][j] = position[j]
                        elif position[j] > self.max[i][j]:
                            self.max[i][j] = position[j]
                        box[j] = self.max[i][j] - self.min[i][j]
                        coverage += box[j] * box[j]
                    coverage = math.sqrt(coverage)
                    self.coverage[i] = coverage
                    string = '{:.3f}'.format(coverage)
                    # if string != coverage_labels[i]:
                    #     print(f'coverage for robot {i}: {string}')
                    coverage_labels[i] = string
                if position[2] < 0.9:  # low position threshold
                    self.ko_count[i] = self.ko_count[i] + 200
                    if self.ko_count[i] > 10000:  # 10 seconds
                        ko = i
                else:
                    self.ko_count[i] = 0
                counter = 10 - self.ko_count[i] // 1000
                string = '' if self.ko_count[i] == 0 else str(counter) if counter > 0 else 'KO'
                if string != ko_labels[i] and string:
                    print(f'robot {i}: {string}')
                ko_labels[i] = string

        return ko



        import psutil        
        CI = os.environ.get("CI")
        wrestler = WrestlerSupervisorMemory()
        loop = 1        
        while True:
            wrestler.simulationReset()
            wrestler.run(CI)
            loop+=1
            print(f'************** loop={loop} *****************')
            for p in psutil.process_iter():
                if 'webots-bin' in p.name() :
                    print(p.name, p.memory_info().rss)    


utilities/robot.py


class RobotAccess(object):
    def __init__(self, robot, mode=1, samplingPeriod=20):
        self.mode=mode
        
        # motors
        self.RShoulderPitchM = robot.getDevice("RShoulderPitch")
        self.LShoulderPitchM = robot.getDevice("LShoulderPitch")
        
        self.LAnklePitchM = robot.getDevice("LAnklePitch")
        self.LKneePitchM = robot.getDevice("LKneePitch")
        self.LHipPitchM = robot.getDevice("LHipPitch")

        self.RAnklePitchM = robot.getDevice("RAnklePitch")
        self.RKneePitchM = robot.getDevice("RKneePitch")
        self.RHipPitchM = robot.getDevice("RHipPitch")

        # sensors
        self.RShoulderPitchS = robot.getDevice("RShoulderPitchS")
        self.LShoulderPitchS = robot.getDevice("LShoulderPitchS")

        self.LAnklePitchS = robot.getDevice("LAnklePitchS")
        self.LKneePitchS = robot.getDevice("LKneePitchS")
        self.LHipPitchS = robot.getDevice("LHipPitchS")

        self.RAnklePitchS = robot.getDevice("RAnklePitchS")
        self.RKneePitchS = robot.getDevice("RKneePitchS")
        self.RHipPitchS = robot.getDevice("RHipPitchS")

        #print('getSamplingPeriod=',self.LAnklePitchS.getSamplingPeriod())
        print(f'samplingPeriod={samplingPeriod}')
        self.LAnklePitchS.enable(samplingPeriod)
        self.LKneePitchS.enable(samplingPeriod)
        self.LHipPitchS.enable(samplingPeriod)
        self.RAnklePitchS.enable(samplingPeriod)
        self.RKneePitchS.enable(samplingPeriod)
        self.RHipPitchS.enable(samplingPeriod)
        robot.step(samplingPeriod)


    def read(self):
        if self.mode == 1:
            return self.readLegs()
        
        if self.mode == 2:
            rsp = self.RShoulderPitchS.getValue()
            lsp = self.LShoulderPitchS.getValue()
            print(f'rsp {rsp} lsp {lsp}')


    def set(self, initial_sensors, actions):        
        if self.mode == 1:
            self.setLegs(initial_sensors, actions)

    def setShoulders(self, left, right):
        self.setMotorPosition(self.LShoulderPitchM,left)           
        self.setMotorPosition(self.RShoulderPitchM,right)           

    def setLegs(self, initial_sensors, actions):
        self.setMotorPosition(self.LHipPitchM,initial_sensors['LHipPitch'] + actions['LHipPitch'])           
        self.setMotorPosition(self.LKneePitchM,initial_sensors['LKneePitch'] + actions['LKneePitch'])
        self.setMotorPosition(self.LAnklePitchM,initial_sensors['LAnklePitch'] + actions['LAnklePitch'])
        self.setMotorPosition(self.RHipPitchM,initial_sensors['RHipPitch'] + actions['RHipPitch'])        
        self.setMotorPosition(self.RKneePitchM,initial_sensors['RKneePitch'] + actions['RKneePitch'])
        self.setMotorPosition(self.RAnklePitchM,initial_sensors['RAnklePitch'] + actions['RAnklePitch'])

    def setMotorPosition(self, motor, position):
        motor.setPosition(min(max(position, motor.min_position),motor.max_position))

    def readLegs(self):
        
        lhp = self.LHipPitchS.getValue()        
        lkp = self.LKneePitchS.getValue()
        lap = self.LAnklePitchS.getValue()
        
        rhp = self.RHipPitchS.getValue()        
        rkp = self.RKneePitchS.getValue()
        rap = self.RAnklePitchS.getValue()
        
        legs = {'LHipPitch': round(lhp, 3), 'LKneePitch': round(lkp,3), 'LAnklePitch': round(lap, 3), 'RHipPitch': round(rhp, 3), 'RKneePitch': round(rkp, 3), 'RAnklePitch': round(rap, 3)}
        
        return legs
```

**Reply by omichel** [reply]
From what you describe, it looks like the memory leak is not in Webots but rather in some shared library used by Webots (maybe in the graphics drivers). That would also explain why you don't get it on all platforms and with depending on the applications started on a machine. I would recommend you to try to upgrade your graphics drivers and see whether it helps.

**Reply by perceptualrobots** [reply]
> That's probably a memory leak in Webots or in the controller library. A way to work around the problem is to modify the referee supervisor so that it [quits](https://cyberbotics.com/doc/reference/supervisor#wb_supervisor_simulation_quit) when done and have a script that launches Webots and repeatedly re-launches Webots as soon as it quits. This way, the memory is cleaned-up regularly and shouldn't grow until Webots crashes.

I am trying this but get error:

`Error: only one Robot instance can be created per controller process. 
`

Am I doing the robot cleanup correctly (with del wrestler)?

```
        while True:
            ex.start_webots()
            wrestler = WrestlerSupervisorServer()
            while True:
                wrestler.simulationReset()
                wrestler.run(port=port)
                ram = ex.webots_ram()
                if ram > ram_limit:
                    logger.info(f'Ram={ram} exceeds limit {ram_limit}, restarting webots')
                    status = 1 #EXIT_SUCCESS
                    wrestler.simulationQuit(status) 
                    del wrestler
                    break
```

**Reply by omichel** [reply]
Yes, you cannot create a Robot object twice in a controller program. Thus you not `del` it and recreate it. This is not supported by the API.

**Reply by perceptualrobots** [reply]
> Yes, you cannot create a Robot object twice in a controller program. Thus you not `del` it and recreate it. This is not supported by the API.

But isn't this what you are suggesting above with the quit and relaunch?

**Reply by omichel** [reply]
No, my suggestion was to quit Webots, not the controller, using the supervisor API. As a consequence, the controller would also quit. That means that Webots should be originally launched by a script that will re-launch Webots whenever it quits (and thus relaunch the robot controllers as well).

**Reply by perceptualrobots** [reply]
> No, my suggestion was to quit Webots, not the controller, using the supervisor API. As a consequence, the controller would also quit. That means that Webots should be originally launched by a script that will re-launch Webots whenever it quits (and thus relaunch the robot controllers as well).

Isn't that what my code above is doing? It starts up Webots as a separate command line process, then if memory is over a certain limit it quits Webots with the supervisor API (wrestler.simulationQuit), then it relaunches Webots and then it recreates the robot controller (which is also a Supervisor). What am I misunderstanding?

**Reply by omichel** [reply]
The launch of Webots should be done in an upper script doing only this:
```python
while True:
    ex.start_webots()
```
Then Webots will automatically open your supervisor controller program (if not declared as `<extern>`) and run it. When the supervisor controller asks Webots to quit, the control will return to your upper script that will start Webots again, etc.

**Reply by perceptualrobots** [reply]
> Then Webots will automatically open your supervisor controller program (if not declared as `<extern>`) and run it. When the supervisor controller asks Webots to quit, the control will return to your upper script that will start Webots again, etc.

My supervisor controller is declared as extern. Does that mean I need to restart the supervisor controller process as well as the Webots process?

The code above is within the participant.py file.

**Reply by omichel** [reply]
> My supervisor controller is declared as extern. Does that mean I need to restart the supervisor controller process as well as the Webots process?

If it is `<extern>` yes, but it would be way simpler to make it not `<extern>`, so that you don't have to care about it and Webots will launch it and terminate it automatically.

**Reply by perceptualrobots** [reply]
> From what you describe, it looks like the memory leak is not in Webots but rather in some shared library used by Webots (maybe in the graphics drivers). That would also explain why you don't get it on all platforms and with depending on the applications started on a machine. I would recommend you to try to upgrade your graphics drivers and see whether it helps.

There was an update pending on the  laptop, so I updated. The desktop had the latest graphics driver. I am also now using another laptop which is only three months old. On all three machines the memory issue happens, even from restart without any other graphics applications running, so it wouldn't appear that the graphics driver is the issue. I will try without extern and see if I can get that to work.

**Comment by ygoumaz** [comment]
To identify the source of the memory leak, you may want to consider using tools such as `Heaptrack` (check out the [brief wiki page](https://github.com/cyberbotics/webots/wiki/Heaptrack) for instructions on how to use it with Webots). It's a user-friendly tool with a sleek interface, which can help you locate memory leaks. Note that `Heaptrack` is compatible only with Linux, and you need to compile Webots in debug mode to use it. Check out [this guide on Linux installation](https://github.com/cyberbotics/webots/wiki/Linux-installation/) to learn how to compile Webots from sources.
